File: Labs/lab5/Dictionary.c
/*	Nicholas Pappas
	nhpappas@ucsc.edu
	ID: 1558554
	Assignment: lab5
*/
//-----------------------------------------------------------------------------
//   Dictionary.c
//   Implementation file for Dictionary ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<ctype.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->key = x;
   N->value = y;
   N->next = NULL;
   return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// DictionaryObj
// Exported reference type
typedef struct DictionaryObj{
   Node head;
   Node tail;
   int numItems;
} DictionaryObj;

typedef struct DictionaryObj* Dictionary;

// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->head = NULL;
   D->tail = NULL;
   D->numItems = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
      if( !isEmpty(*pD) ) makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numItems==0);
}

// size()
// pre: none
// post: returns the number of entries in this Dictionary
int size(Dictionary D) {
	return D->numItems;
	}

// findKey()
// returns a reference to the Node with this key in this Dictionary
Node findKey(Dictionary D, char* k){
    Node N;
    if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling findKey() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
	}
    N = D->head;
    while(N != NULL){
		if(strcmp(k, N->key)==0){
			return N;}
		else{
			N = N->next;}
		}
	return NULL;
	}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
   Node N;
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
    }
   if(findKey(D, k)!= NULL){
	  N = findKey(D, k);
	  return N->value;}
   else{
	  return NULL;
	}
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
   Node N;
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( lookup(D,k)!=NULL ){
      fprintf(stderr, "insert(): cannot insert duplicate keys\n");
      exit(EXIT_FAILURE);
   }
   
   if( D->head == NULL ){
	 N = newNode(k,v);
	 D->head = N;
	 D->tail = N;
	 }
   else if( D->head == D->tail ){
	 N = D->head;
	 N->next = newNode(k,v);
	 D->tail = N->next;}
   else{
	 N = D->tail;
	 N->next = newNode(k,v);
	 D->tail = N->next;
    } 
	 D->numItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
   Node P;
   Node S;
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( D->numItems==0 ){
      fprintf(stderr, "Dictionary Error: calling delete() on empty Dictionary\n");
      exit(EXIT_FAILURE);
   }
   
   if( lookup(D,k)==NULL ){
      fprintf(stderr, "delete(): cannot delete non-existent key \n");
      exit(EXIT_FAILURE);
   }
   else if( findKey(D,k)== D->head ){ //delete head if only node
	  if( D->head == D->tail ){
		  P = D->head;
		  free(P);
		  D->head = NULL;
		  D->tail = NULL;}
		 else{
          P = D->head;				//delete head if multiple nodes
          D->head = P->next;
		  free(P);
		  }
	  }
	  else if(findKey(D,k)== D->tail){  //delete tail
		 P = D->head;
		 while(P->next != D->tail){
			 P = P->next;}
		 D->tail = P;
		 P = P->next;
		 P = NULL;
		 free(P);
	  }else{
         S = findKey(D,k);  //delete middle
         P = D->head;
		 while(P->next != S){
			 P = P->next;}
         P->next = S->next;
		 free(S);
      }
   D->numItems--;
   
}

// makeEmpty()
// resets D to the empty state
// pre: !isEmpty(D)
void makeEmpty(Dictionary D){
	Node N;
	Node temp;
	if(D==NULL)
	{
		fprintf(stderr,"Dictionary Error: calling makeEmpty() on NULL Dictionary reference");
		exit(EXIT_FAILURE);
	}
	N=D->head;
	while(N!=NULL)
	{
		temp=N->next;
		free(N);
		N=temp;
	}
	D->numItems=0;
}

// printDictionary()
// prints a text representation of D to the file pointed to by out
// pre: none
void printDictionary(FILE* out, Dictionary D){
   Node N;
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   for(N=D->head; N!=NULL; N=N->next) fprintf(out, "%s %s\n", N->key, N->value );
}

File: Labs/lab7/Dictionary.java
/*	Nicholas Pappas
	nhpappas@ucsc.edu
	ID: 1558554
	Assignment: lab7
*/
//-----------------------------------------------------------------------------
// Dictionary.java
// Linked List implementation of the Dictionary BST ADT
//-----------------------------------------------------------------------------

//import java.io.*;
//import java.util.Scanner;

public class Dictionary implements DictionaryInterface {

//PrintWriter out = null;
//out = new PrintWriter(new FileWriter(args[1]));

   // private inner Node class
   private class Node {
      String key, value;
      Node left, right;

      Node(String x, String y){
         key = x;
		 value = y;
         left = right = null;
      }
   }

   // Fields for the Dictionary class
   private Node root;     // reference to first Node in tree
   private int numPairs;  // number of items in this Dictionary

   // Dictionary()
   // constructor for the Dictionary class
   public Dictionary(){
      root = null;
      numPairs = 0;
   }
   
   // findKey()
   // returns a reference to the Node with this key in this Dictionary
   private Node findKey(Node R, String key){
      if(R==null || key.compareTo(R.key)==0){
		  return R;
	  }
	  if(key.compareTo(R.key)<0){
		  return findKey(R.left, key);
	  }
	  else{ //(key.compareTo(R.key)<0)
		  return findKey(R.right, key);
	  }
   }
   
	// findParent()
	// returns the parent of N in the subtree rooted at R, or returns null 
	// if N is equal to R. (pre: R != null)
   private Node findParent(Node N, Node R){
	Node P=null;
	if( N!=R ){
		P = R;
		while( P.left!=N && P.right!=N ){
			if(N.key.compareTo(P.key)<0){
				P = P.left;}
			else{
				P = P.right;
				}
			}
		}
	return P;
   }
   
   // findLeftmost()
	// returns the leftmost Node in the subtree rooted at R, or null if R is null.
	private Node findLeftmost(Node R){
	Node L = R;
	if( L!=null ){ 
		for( ; L.left!=null; L=L.left){}
		}
		return L;
	}
	
	// printInOrder()
	// prints the (key, value) pairs belonging to the subtree rooted at R in order
	// of increasing keys to file pointed to by out.
	private String printInOrder(Node R, String s){
	if( R!=null ){
		s = printInOrder(R.left, s);
		s += R.key + " " + R.value + "\n";
		s = printInOrder(R.right, s);
		} else{}
	return s;
	}

	// deleteAll()
	// deletes all Nodes in the subtree rooted at N.
	void deleteAll(Node N){
	if( N!=null ){
		deleteAll(N.left);
		deleteAll(N.right);
		N = null;
		}
	}



   // Implementation of ADT operations //////////////////////////////////////

   // isEmpty()
   // pre: none
   // post: returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty(){
      return(numPairs == 0);
   }

   // size()
   // pre: none
   // post: returns the number of entries in this Dictionary
   public int size() {
      return numPairs;
   }

   // lookup()
   // pre: none
   // returns value associated key, or null reference if no such key exists
   public String lookup(String key) {
      Node N = findKey(root, key);
	  if(N == null){
		return null;
	  }
      return N.value;
   }

   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   public void insert(String key, String value) throws DuplicateKeyException{
	Node N, A, B;
      if(findKey(root, key)!=null){
         throw new DuplicateKeyException(
            "insert(): cannot insert duplicate keys " + key);
      }
		N = new Node(key, value);
		B = null;
		A = root;
		while( A!=null ){
			B = A;
			if( key.compareTo(A.key)<0 ){ A = A.left;}
			else {A = A.right;}
		}
		if( B==null ){ root = N;}
		else if( key.compareTo(B.key)<0 ){ B.left = N;}
		else{ B.right = N;}
		numPairs++;
   }

   // delete()
   // deletes pair with the given key
   // pre: lookup(key)!=null
   public void delete(String key) throws KeyNotFoundException{
      Node N, P, S;
   N = findKey(root, key);
   if(findKey(root, key)==null){
         throw new KeyNotFoundException(
            "delete(): cannot delete non-existent key " + key);
      }
   if( N.left==null && N.right==null ){  // case 1 (no children)
      if( N==root ){
         root = null;
      }else{
         P = findParent(N, root);
         if( P.right==N ){ P.right = null;}
         else{ P.left = null;}
      }
   }else if( N.right==null ){  // case 2 (left but no right child)
      if( N==root ){
         root = N.left;
		 N= null;
      }else{
         P = findParent(N, root);
         if( P.right==N ){ P.right = N.left;}
         else{ P.left = N.left;}
		 N= null;
      }
   }else if( N.left==null ){  // case 2 (right but no left child)
      if( N==root ){
         root = N.right;
         N=null;
      }else{
         P = findParent(N, root);
         if( P.right==N ) P.right = N.right;
         else P.left = N.right;
         N=null;
      }
   }else{                     // case3: (two children: N.left!=null && N.right!=null)
      S = findLeftmost(N.right);
      N.key = S.key;
      N.value = S.value;
      P = findParent(S, N);
      if( P.right==S ){ P.right = S.right;}
      else{ P.left = S.right;}
      S= null;
   }
   numPairs--;
   }

   // makeEmpty()
   // pre: none
   public void makeEmpty(){
    deleteAll(root);
	root = null;
	numPairs = 0;
   }

   // toString
   // pre: none
   // post:  prints current state to stdout
   // Overrides Object's toString() method
   public String toString(){
      String s = printInOrder(root, "");
      return s;
   }
}